neo/IO/Caching/DataCache.cs

```diff
@@ -78,137 +84,159 @@ public void Commit()
-                    trackable.State = TrackState.Changed;
+                    trackable = new Trackable
+                    {
+                        Key = key,
+                        Item = TryGetInternal(key)
+                    };
+                    if (trackable.Item == null)
+                    {
+                        if (factory == null) throw new KeyNotFoundException();
+                        trackable.Item = factory();
+                        trackable.State = TrackState.Added;
+                    }
+                    else
+                    {
+                        trackable.State = TrackState.Changed;
+                    }
+                    dictionary.Add(key, trackable);
                 }
-                dictionary.Add(key, trackable);
+                return trackable.Item;
             }
-            return trackable.Item;
         }
 
         public TValue GetOrAdd(TKey key, Func<TValue> factory)
         {
-            if (dictionary.TryGetValue(key, out Trackable trackable))
-            {
-                if (trackable.State == TrackState.Deleted)
-                {
-                    trackable.Item = factory();
-                    trackable.State = TrackState.Changed;
-                }
-            }
-            else
+            lock (dictionary)
             {
-                trackable = new Trackable
-                {
-                    Key = key,
-                    Item = TryGetInternal(key)
-                };
-                if (trackable.Item == null)
+                if (dictionary.TryGetValue(key, out Trackable trackable))
                 {
-                    trackable.Item = factory();
-                    trackable.State = TrackState.Added;
+                    if (trackable.State == TrackState.Deleted)
+                    {
+                        trackable.Item = factory();
+                        trackable.State = TrackState.Changed;
+                    }
                 }
                 else
                 {
-                    trackable.State = TrackState.None;
+                    trackable = new Trackable
+                    {
+                        Key = key,
+                        Item = TryGetInternal(key)
+                    };
+                    if (trackable.Item == null)
+                    {
+                        trackable.Item = factory();
+                        trackable.State = TrackState.Added;
+                    }
+                    else
+                    {
+                        trackable.State = TrackState.None;
+                    }
+                    dictionary.Add(key, trackable);
                 }
-                dictionary.Add(key, trackable);
+                return trackable.Item;
             }
-            return trackable.Item;
         }
 
         public TValue TryGet(TKey key)
         {
-            if (dictionary.TryGetValue(key, out Trackable trackable))
+            lock (dictionary)
```

Comment:
Specifically multiple readers calling this should not block one another.

Reply:
I have tested this class with R/W Locks but no performance improvement. Because although it is the `TryGet` method, it has to modify the dictionary internally.

Reply:
I see; I took for granted that this method modifies the dictionary.